xcp_d/utils/utils.py

```diff
@@ -380,108 +392,106 @@ def denoise_with_nilearn(
-        # but overwrite the affected volumes later.
-        interpolated_unfiltered_bold = signal._interpolate_volumes(
-            interpolated_unfiltered_bold,
-            sample_mask=sample_mask,
-            t_r=TR,
-            extrapolate=True,
-        )
 
-        # Replace any high-motion volumes at the beginning or end of the run with the closest
-        # low-motion volume's data.
-        # Use https://stackoverflow.com/a/48106843/2589328 to group consecutive blocks of outliers.
-        gaps = [[s, e] for s, e in zip(outlier_idx, outlier_idx[1:]) if s + 1 < e]
-        edges = iter(outlier_idx[:1] + sum(gaps, []) + outlier_idx[-1:])
-        consecutive_outliers_idx = list(zip(edges, edges))
-        first_outliers = consecutive_outliers_idx[0]
-        last_outliers = consecutive_outliers_idx[-1]
-
-        # Replace outliers at beginning of run
-        if first_outliers[0] == 0:
-            LOGGER.warning(
-                f"Outlier volumes at beginning of run ({first_outliers[0]}-{first_outliers[1]}) "
-                "will be replaced with first non-outlier volume's values."
-            )
-            interpolated_unfiltered_bold[: first_outliers[1] + 1, :] = (
-                interpolated_unfiltered_bold[first_outliers[1] + 1, :]
-            )
+def _interpolate(*, arr, sample_mask, TR):
+    from nilearn import signal
 
-        # Replace outliers at end of run
-        if last_outliers[1] == n_volumes - 1:
-            LOGGER.warning(
-                f"Outlier volumes at end of run ({last_outliers[0]}-{last_outliers[1]}) "
-                "will be replaced with last non-outlier volume's values."
-            )
-            interpolated_unfiltered_bold[last_outliers[0] :, :] = interpolated_unfiltered_bold[
-                last_outliers[0] - 1, :
-            ]
+    outlier_idx = list(np.where(~sample_mask)[0])
+    n_volumes = arr.shape[0]
 
-    else:
-        interpolated_unfiltered_bold = censored_denoised_bold
-
-    # Now apply the bandpass filter to the interpolated, denoised data
-    if low_pass is not None and high_pass is not None:
-        # TODO: Replace with nilearn.signal.butterworth once 0.10.1 is released.
-        interpolated_filtered_bold = butter_bandpass(
-            interpolated_unfiltered_bold,
-            sampling_rate=1 / TR,
-            low_pass=low_pass,
-            high_pass=high_pass,
-            order=filter_order / 2,
-            padtype="constant",
-            padlen=n_volumes - 1,
+    interpolated_arr = signal._interpolate_volumes(
+        arr,
+        sample_mask=sample_mask,
+        t_r=TR,
+        extrapolate=True,
+    )
+    # Replace any high-motion volumes at the beginning or end of the run with the closest
+    # low-motion volume's data.
+    # Use https://stackoverflow.com/a/48106843/2589328 to group consecutive blocks of outliers.
+    gaps = [[s, e] for s, e in zip(outlier_idx, outlier_idx[1:]) if s + 1 < e]
```

Comment:
Sorry, what are s and e again?

Reply:
I believe they stand for `start` and `end`. I will update the variable names to make it clearer.